### PR TITLE
tty/usbacm: fix deadlock on device removal

### PR DESCRIPTION
**Pull request by ziemleszcz**
JIRA: RTOS-426

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/usbacm/usbacm.c [hunk 1]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
 }
 
 
-static void _usbacm_put(usbacm_dev_t *dev)
+static void usbacm_remove(usbacm_dev_t *dev)
```
Comment (nalajcie) [inline tty/usbacm/usbacm.c:199]: I would name it `_free` to highlight that `dev` becomes invalid after calling this function

File: tty/usbacm/usbacm.c [hunk 2]
```diff
@@ -694,8 +725,10 @@ static void usbthr(void *arg)
 				break;
 			case usb_msg_deletion:
 				mutexLock(usbacm_common.lock);
-				_usbacm_handleDeletion(&umsg->deletion);
+				_usbacm_handleDeletion(&umsg->deletion, &devices_to_remove);
 				mutexUnlock(usbacm_common.lock);
+				usbacm_remove_all(devices_to_remove);
+				devices_to_remove = NULL;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:731]: Imho it would be better for `usbacm_remove_all` to NULLify the list

File: tty/usbacm/usbacm.c [hunk 3]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
...
+		free(dev->fifo);
+		dev->fifo = NULL;
+		resourceDestroy(dev->rxLock);
+		resourceDestroy(dev->rxCond);
+	}
+	free(dev);
+}
+
+
+static void usbacm_remove_all(usbacm_dev_t *devices)
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:213]: Use camelCase
```suggestion
static void usbacm_removeAll(usbacm_dev_t *devices)
```

File: tty/usbacm/usbacm.c [hunk 4]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
...
 
 
-static void _usbacm_put(usbacm_dev_t *dev)
+static void usbacm_remove(usbacm_dev_t *dev)
+{
+	fprintf(stdout, "usbacm: Device removed: %s\n", dev->path);
+	remove(dev->path);
+	if (dev->fifo != NULL) {
+		free(dev->fifo);
+		dev->fifo = NULL;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:205]: Seems redundant as we will `free` `dev` anyway

**Review by nalajcie — COMMENTED** [review]
Can you describe the original deadlock? I see the `_usbacm_handleDeletion` was called with `usbacm_common.lock` taken but I don see where we would try to take it again...

File: tty/usbacm/usbacm.c [hunk 5]
```diff
@@ -683,6 +713,7 @@ static void usbthr(void *arg)
 	msg_t msg;
 	usb_msg_t *umsg = (usb_msg_t *)msg.i.raw;
 	unsigned long rid;
+	usbacm_dev_t *devices_to_remove = NULL;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:716]:
```suggestion
	usbacm_dev_t *devicesToRemove = NULL;
```

File: tty/usbacm/usbacm.c [hunk 6]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
...
 
 
 static void usbacm_put(usbacm_dev_t *dev)
 {
 	mutexLock(usbacm_common.lock);
-	_usbacm_put(dev);
+	dev = _usbacm_put(dev);
 	mutexUnlock(usbacm_common.lock);
+
+	if (dev) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:246]: Nitpick
```suggestion
	if (dev != NULL) {
```

File: tty/usbacm/usbacm.c [hunk 7]
```diff
@@ -662,12 +692,12 @@ static int _usbacm_handleDeletion(usb_deletion_t *del)
...
 		if (dev->instance.bus == del->bus && dev->instance.dev == del->dev &&
 				dev->instance.interface == del->interface) {
-			fprintf(stdout, "usbacm: Device removed: %s\n", dev->path);
 			if (dev == next)
 				cont = 0;
-
-			remove(dev->path);
-			_usbacm_put(dev);
+			dev = _usbacm_put(dev);
+			if (dev) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:698]: Nitpick
```suggestion
			if (_usbacm_put(dev) != NULL) {
```

File: tty/usbacm/usbacm.c [hunk 8]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
...
+
+	if (dev == NULL) {
+		return;
+	}
+
+	do {
+		next = dev->next;
+		usbacm_remove(dev);
+		dev = next;
+	} while (dev != devices);
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:225]: Nitpick
```suggestion
	if (dev != NULL) {
		do {
			next = dev->next;
			usbacm_remove(dev);
			dev = next;
		} while (dev != devices);
	}
```

File: tty/usbacm/usbacm.c [hunk 9]
```diff
@@ -196,26 +196,56 @@ static usbacm_dev_t *usbacm_get(int id)
...
+}
+
+
+static usbacm_dev_t *_usbacm_put(usbacm_dev_t *dev)
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:229]: Just a suggestion: I think we can also return simply number of references left. We don't really need a pointer as a retval here.

**Comment by agkaminski** [comment]
@nalajcie part of `remove()` is sending a msg to the server of a removed file - us in this case. So we send a msg to ourself and can cause a deadlock this way

**Comment by nalajcie** [comment]
> @nalajcie part of `remove()` is sending a msg to the server of a removed file - us in this case. So we send a msg to ourself and can cause a deadlock this way

Ahh, ok, that's what I'm missing. On Linux You can easily remove and recreate the device without side effects (only device major/minor is important). Thx for the summary.

File: tty/usbacm/usbacm.c [hunk 10]
```diff
@@ -196,26 +196,57 @@ static usbacm_dev_t *usbacm_get(int id)
...
 	if (--dev->rfcnt == 0) {
-		if (dev->fifo != NULL) {
-			free(dev->fifo);
-			dev->fifo = NULL;
-			resourceDestroy(dev->rxLock);
-			resourceDestroy(dev->rxCond);
-		}
 		LIST_REMOVE(&usbacm_common.devices, dev);
-		free(dev);
+		return 0;
```
Comment (kemonats) [inline tty/usbacm/usbacm.c:232]:
```suggestion
```
Unnecessary, `dev->rfcnt == 0`.

**Pull request merged**